### ISSUE TITLE
Cuda issues.

### ISSUE THREAD

**Issue by anonymau5** [nagadomi/waifu2x#110]
Installed with the latest instructions and received the following.  It seems that it is having trouble accessing CUDA although it is installed.

/usr/local/bin/luajit: /usr/local/share/lua/5.1/trepl/init.lua:383: lib/w2nn.lua:18: Failed to load CUDA modules. Please check the CUDA Settings.                               
stack traceback:                                                                        
        [C]: in function 'error'                                                        
        /usr/local/share/lua/5.1/trepl/init.lua:383: in function 'require'              
        web.lua:7: in main chunk                                                        
        [C]: in function 'dofile'                                                       
        /usr/local/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk             
        [C]: at 0x00404270

**Comment by nagadomi** [comment]
I added more informative error message at https://github.com/nagadomi/waifu2x/commit/958dcfaeb8dfd3c71df65282168b211f745ff970.
Could you retry and post the error message?

**Comment by anonymau5** [comment]
A little more information on this machine;

This machine uses a 550ti and has been running Waifu2x just fine for some time.  I recently had a driver issue and had to purge and reinstall the nvidia drivers.  The latest 352.63 drivers along with CUDA 7..5.18 is installed.  There may be remnants of the older install still there.  


/usr/local/bin/luajit: /usr/local/share/lua/5.1/trepl/init.lua:383: lib/w2nn.lua:18: Failed to load CUDA modules. Please check the CUDA Settings.
---
/usr/local/share/lua/5.1/trepl/init.lua:383: /usr/local/share/lua/5.1/cutorch/init.lua:2: cannot load '/usr/local/lib/lua/5.1/libcutorch.so'
stack traceback:
        [C]: in function 'error'
        /usr/local/share/lua/5.1/trepl/init.lua:383: in function 'require'
        web.lua:7: in main chunk
        [C]: in function 'dofile'
        /usr/local/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
        [C]: at 0x00404270

**Comment by anonymau5** [comment]
user@Mami:~/.Waifu2X$ sudo apt-get install libsnappy-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
libsnappy-dev is already the newest version.
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.

user@Mami:~/.Waifu2X$ sudo apt-get install libgraphicsmagick-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Note, selecting 'libgraphicsmagick1-dev' instead of 'libgraphicsmagick-dev'
libgraphicsmagick1-dev is already the newest version.
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.

user@Mami:~/.Waifu2X$ sudo luarocks install graphicsmagick
Installing https://raw.githubusercontent.com/torch/rocks/master/graphicsmagick-1.scm-0.rockspec...
Using https://raw.githubusercontent.com/torch/rocks/master/graphicsmagick-1.scm-0.rockspec... switching to 'build' mode
Cloning into 'graphicsmagick'...
remote: Counting objects: 29, done.
remote: Compressing objects: 100% (29/29), done.
remote: Total 29 (delta 0), reused 19 (delta 0), pack-reused 0
Receiving objects: 100% (29/29), 788.37 KiB | 0 bytes/s, done.
Checking connectivity... done.
Updating manifest for /usr/local/lib/luarocks/rocks
graphicsmagick 1.scm-0 is now built and installed in /usr/local/ (license: BSD)

user@Mami:~/.Waifu2X$ sudo luarocks install lua-csnappy
Installing https://raw.githubusercontent.com/rocks-moonscript-org/moonrocks-mirror/master/lua-csnappy-0.1.3-1.src.rock...
Using https://raw.githubusercontent.com/rocks-moonscript-org/moonrocks-mirror/master/lua-csnappy-0.1.3-1.src.rock... switching to 'build' mode
gcc -O2 -fPIC -I/usr/local/include -c lsnappy.c -o lsnappy.o
gcc -shared -o snappy.so -L/usr/local/lib lsnappy.o
Updating manifest for /usr/local/lib/luarocks/rocks
lua-csnappy 0.1.3-1 is now built and installed in /usr/local/ (license: BSD)

user@Mami:~/.Waifu2X$ sudo luarocks install md5
Installing https://raw.githubusercontent.com/rocks-moonscript-org/moonrocks-mirror/master/md5-1.2-1.src.rock...
Using https://raw.githubusercontent.com/rocks-moonscript-org/moonrocks-mirror/master/md5-1.2-1.src.rock... switching to 'build' mode
gcc -O2 -fPIC -I/usr/local/include -c src/compat-5.2.c -o src/compat-5.2.o -Isrc/
gcc -O2 -fPIC -I/usr/local/include -c src/md5.c -o src/md5.o -Isrc/
gcc -O2 -fPIC -I/usr/local/include -c src/md5lib.c -o src/md5lib.o -Isrc/
gcc -shared -o md5/core.so -L/usr/local/lib src/compat-5.2.o src/md5.o src/md5lib.o
gcc -O2 -fPIC -I/usr/local/include -c src/compat-5.2.c -o src/compat-5.2.o -Isrc/
gcc -O2 -fPIC -I/usr/local/include -c src/des56.c -o src/des56.o -Isrc/
gcc -O2 -fPIC -I/usr/local/include -c src/ldes56.c -o src/ldes56.o -Isrc/
gcc -shared -o des56.so -L/usr/local/lib src/compat-5.2.o src/des56.o src/ldes56.o
Updating manifest for /usr/local/lib/luarocks/rocks
md5 1.2-1 is now built and installed in /usr/local/ (license: MIT/X11)

user@Mami:~/.Waifu2X$ sudo luarocks install uuid
Installing https://raw.githubusercontent.com/rocks-moonscript-org/moonrocks-mirror/master/uuid-0.2-1.src.rock...
Using https://raw.githubusercontent.com/rocks-moonscript-org/moonrocks-mirror/master/uuid-0.2-1.src.rock... switching to 'build' mode
Updating manifest for /usr/local/lib/luarocks/rocks
uuid 0.2-1 is now built and installed in /usr/local/ (license: Apache 2.0)

user@Mami:~/.Waifu2X$ sudo PREFIX=$HOME/torch/install luarocks install turbo
Installing https://raw.githubusercontent.com/rocks-moonscript-org/moonrocks-mirror/master/turbo-2.0-4.rockspec...
Using https://raw.githubusercontent.com/rocks-moonscript-org/moonrocks-mirror/master/turbo-2.0-4.rockspec... switching to 'build' mode
Cloning into 'turbo'...
remote: Counting objects: 165, done.
remote: Compressing objects: 100% (145/145), done.
remote: Total 165 (delta 28), reused 67 (delta 16), pack-reused 0
Receiving objects: 100% (165/165), 645.57 KiB | 0 bytes/s, done.
Resolving deltas: 100% (28/28), done.
Checking connectivity... done.
Note: checking out 'f43bffa558742548b4bba1f39ef2771d97d7041c'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b new_branch_name

Warning: variable CFLAGS was not passed in build_variables
make -C deps/http-parser library
make[1]: Entering directory `/tmp/luarocks_turbo-2.0-4-6086/turbo/deps/http-parser'
gcc -I. -DHTTP_PARSER_STRICT=0  -fPIC -Wall -Wextra -Werror -O3  -c http_parser.c -o libhttp_parser.o
gcc  -shared -Wl,-soname=libhttp_parser.so.2.1 -o libhttp_parser.so.2.1 libhttp_parser.o
make[1]: Leaving directory `/tmp/luarocks_turbo-2.0-4-6086/turbo/deps/http-parser'
gcc -Ideps/http-parser/ -shared -O3 -Wall -g -fPIC deps/http-parser/libhttp_parser.o deps/turbo_ffi_wrap.c -o libtffi_wrap.so -lcrypto -lssl
==== Installing Turbo.lua v2.0.4 to: ====
==== /home/user/torch/install/lib/lua/5.1 and ====
==== /home/user/torch/install/share/luajit-2.0.4 ====
mkdir -p /home/user/torch/install/lib
mkdir -p /home/user/torch/install/bin
mkdir -p /home/user/torch/install/share/lua/5.1/turbo
mkdir -p /home/user/torch/install/share/luajit-2.0.4/turbo
cp -r turbo/* /home/user/torch/install/share/lua/5.1/turbo
cp -r turbo.lua /home/user/torch/install/share/lua/5.1
cp -r turbovisor.lua /home/user/torch/install/share/lua/5.1
cp -r turbo/* /home/user/torch/install/share/luajit-2.0.4/turbo
cp -r turbo.lua /home/user/torch/install/share/luajit-2.0.4
cp -r turbovisor.lua /home/user/torch/install/share/luajit-2.0.4
install -m 0755 bin/turbovisor /home/user/torch/install/bin
==== Building 3rdparty modules ====
make -C deps/http-parser library
make[1]: Entering directory `/tmp/luarocks_turbo-2.0-4-6086/turbo/deps/http-parser'
gcc  -shared -Wl,-soname=libhttp_parser.so.2.1 -o libhttp_parser.so.2.1 libhttp_parser.o
make[1]: Leaving directory `/tmp/luarocks_turbo-2.0-4-6086/turbo/deps/http-parser'
gcc -Ideps/http-parser/ -shared -O3 -Wall -g -fPIC deps/http-parser/libhttp_parser.o deps/turbo_ffi_wrap.c -o libtffi_wrap.so -lcrypto -lssl
==== Installing libtffi_wrap ====
test -f libtffi_wrap.so && \
        install -m 0755 libtffi_wrap.so /home/user/torch/install/lib/libtffi_wrap.so.2.0.4 && \
        ldconfig -n /home/user/torch/install/lib && \
        ln -sf libtffi_wrap.so.2.0.4 /home/user/torch/install/lib/libtffi_wrap.so
==== Successfully installed Turbo.lua 2.0.4 to /home/user/torch/install ====
Updating manifest for /usr/local/lib/luarocks/rocks
turbo 2.0-4 is now built and installed in /usr/local/ (license: Apache 2.0)

user@Mami:~/.Waifu2X$ th waifu2x.lua
/usr/local/bin/luajit: /usr/local/share/lua/5.1/trepl/init.lua:383: lib/w2nn.lua:18: Failed to load CUDA modules. Please check the CUDA Settings.
---
/usr/local/share/lua/5.1/trepl/init.lua:383: /usr/local/share/lua/5.1/cutorch/init.lua:2: cannot load '/usr/local/lib/lua/5.1/libcutorch.so'
stack traceback:
        [C]: in function 'error'
        /usr/local/share/lua/5.1/trepl/init.lua:383: in function 'require'
        waifu2x.lua:5: in main chunk
        [C]: in function 'dofile'
        /usr/local/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
        [C]: at 0x00404270

**Comment by nagadomi** [comment]
It seems to be a cutorch(cuda module of torch data structure) issue.
Please try to reinstall that.
```
luarocks cutorch
luarocks cunn
```

**Comment by anonymau5** [comment]
user@Mami:~/.Waifu2X$ luarocks cutorch

Error: Unknown command: cutorch
user@Mami:~/.Waifu2X$ luarocks cunn

Error: Unknown command: cunn
user@Mami:~/.Waifu2X$

**Comment by anonymau5** [comment]
Ah, just noticed I need to do a sudo luarocks install cutorch  Now I'm getting this.

user@Mami:~/.Waifu2X$ sudo luarocks install cutorch
Installing https://raw.githubusercontent.com/torch/rocks/master/cutorch-scm-1.rockspec...
Using https://raw.githubusercontent.com/torch/rocks/master/cutorch-scm-1.rockspec... switching to 'build' mode
Cloning into 'cutorch'...
remote: Counting objects: 107, done.
remote: Compressing objects: 100% (104/104), done.
remote: Total 107 (delta 8), reused 34 (delta 1), pack-reused 0
Receiving objects: 100% (107/107), 156.81 KiB | 0 bytes/s, done.
Resolving deltas: 100% (8/8), done.
Checking connectivity... done.
cmake -E make_directory build && cd build && cmake .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH="/usr/local/bin/.." -DCMAKE_INSTALL_PREFIX="/usr/local/lib/luarocks/rocks/cutorch/scm-1" && make -j$(getconf _NPROCESSORS_ONLN) install

-- The C compiler identification is GNU 4.8.4
-- The CXX compiler identification is GNU 4.8.4
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Found Torch7 in /usr/local
CMake Error at /usr/share/cmake-2.8/Modules/FindCUDA.cmake:548 (message):
  Specify CUDA_TOOLKIT_ROOT_DIR
Call Stack (most recent call first):
  CMakeLists.txt:7 (FIND_PACKAGE)


-- Configuring incomplete, errors occurred!
See also "/tmp/luarocks_cutorch-scm-1-6154/cutorch/build/CMakeFiles/CMakeOutput.log".

Error: Build error: Failed building.
user@Mami:~/.Waifu2X$ sudo luarocks install cunn
Installing https://raw.githubusercontent.com/torch/rocks/master/cunn-scm-1.rockspec...
Using https://raw.githubusercontent.com/torch/rocks/master/cunn-scm-1.rockspec... switching to 'build' mode
Cloning into 'cunn'...
remote: Counting objects: 76, done.
remote: Compressing objects: 100% (70/70), done.
remote: Total 76 (delta 20), reused 17 (delta 2), pack-reused 0
Receiving objects: 100% (76/76), 93.95 KiB | 0 bytes/s, done.
Resolving deltas: 100% (20/20), done.
Checking connectivity... done.
cmake -E make_directory build && cd build && cmake .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH="/usr/local/bin/.." -DCMAKE_INSTALL_PREFIX="/usr/local/lib/luarocks/rocks/cunn/scm-1" && make -j$(getconf _NPROCESSORS_ONLN) install

-- The C compiler identification is GNU 4.8.4
-- The CXX compiler identification is GNU 4.8.4
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Found Torch7 in /usr/local
CMake Error at /usr/share/cmake-2.8/Modules/FindCUDA.cmake:548 (message):
  Specify CUDA_TOOLKIT_ROOT_DIR
Call Stack (most recent call first):
  CMakeLists.txt:7 (FIND_PACKAGE)


-- Configuring incomplete, errors occurred!
See also "/tmp/luarocks_cunn-scm-1-480/cunn/build/CMakeFiles/CMakeOutput.log".

Error: Build error: Failed building.

**Comment by nagadomi** [comment]
Sorry, `luarocks install cutorch` and `luarocks install cunn` are correct.
If you use `sudo`, Maybe root does not read environment variables from `user`.
```
sudo su
export PATH=/usr/local/cuda/bin:$PATH
export LD_LIBRARY_PATH=/usr/local/cuda/lib64:$LD_LIBRARY_PATH
export CUDA_BIN_PATH=/usr/local/cuda/bin
luarocks install cutorch
luarocks install cunn
```

I recommend you to reinstall torch with non root environment.
See http://torch.ch/docs/getting-started.html
Unstall command: https://github.com/torch/ezinstall#updating-from-a-previous-version

**Comment by anonymau5** [comment]
Reinstalled torch with non root environment.  Torch seems to have installed fine.  No errors.

user@Mami:~/.Waifu2X$ th web.lua
/home/user/torch/install/bin/luajit: /home/user/torch/install/share/lua/5.1/trepl/init.lua:384: lib/w2nn.lua:18: Failed to load CUDA modules. Please check the CUDA Settings.
---
/home/user/torch/install/share/lua/5.1/trepl/init.lua:384: module 'cutorch' not found:No LuaRocks module found for cutorch
        no field package.preload['cutorch']
        no file 'lib/cutorch.lua'
        no file '/home/user/.luarocks/share/lua/5.1/cutorch.lua'
        no file '/home/user/.luarocks/share/lua/5.1/cutorch/init.lua'
        no file '/home/user/torch/install/share/lua/5.1/cutorch.lua'
        no file '/home/user/torch/install/share/lua/5.1/cutorch/init.lua'
        no file './cutorch.lua'
        no file '/home/user/torch/install/share/luajit-2.1.0-beta1/cutorch.lua'
        no file '/usr/local/share/lua/5.1/cutorch.lua'
        no file '/usr/local/share/lua/5.1/cutorch/init.lua'
        no file '/home/user/.luarocks/lib/lua/5.1/cutorch.so'
        no file '/home/user/torch/install/lib/lua/5.1/cutorch.so'
        no file '/home/user/torch/install/lib/cutorch.so'
        no file './cutorch.so'
        no file '/usr/local/lib/lua/5.1/cutorch.so'
        no file '/usr/local/lib/lua/5.1/loadall.so'
stack traceback:
        [C]: in function 'error'
        /home/user/torch/install/share/lua/5.1/trepl/init.lua:384: in function 'require'
        web.lua:7: in main chunk
        [C]: in function 'dofile'
        ...user/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
        [C]: at 0x00406670

**Comment by anonymau5** [comment]
I'm not sure why this newer version failed but I've rolled back to an older version with success.

**Comment by nagadomi** [comment]
It seems that cutorch is still not installed. I guess that CUDA is not detected from torch installer.
Is there nvcc?
```
nvcc --version
```
Then install cutorch
```
luarocks install cutorch
luarocks install cunn
```

**Comment by anonymau5** [comment]
I got an older version up and going.  I'll copy the newer version and see if it'll run.

**Comment by liamquin** [comment]
I found that FindCUDA didn't work if nvcc is in /bin - I had to put /usr/bin first in my PATH. This is because FindCUDA strips /bin/ from the directory when determining CUDA_HOME and if it's /bin then nothing is left.

I got everything built but then it wouldn't run; looks like my nvidia card is too old, as cutorch gives me Unsupported gpu architecture 'compute_52' - I gave up for now.

**Comment by nagadomi** [comment]
>nvcc fatal   : Unsupported gpu architecture 'compute_52'

I don't know your GPU model but this error is caused by old nvcc(CUDA 6.5?). 
CUDA 7.5 is required.

**Comment by liamquin** [comment]
Thank you for replying. Yes, 6.5, since 7.5 doesn't support my GPU  - NVIDIA Corporation GT216GLM [Quadro FX 880M]. I did find a version that is patched (by NVidia) to support it, but it's a 1GByte download and unfortunately it also includes the kernel drivers, which will cause other problems on my system. So I gave up for now, will try on a different machine when I get the chance. I'd love to see waifu2x written as a GEGL operation by the way.

Bu fixing FindNVCC to work if nvcc is in /bin and not /usr/bin might be worthwhile as Linux systems are moving to have /usr/bin just be a symlink to /bin.
